[Feat] 투표 초기화 기능 추가

Nine-JH opened the pull request:
## 변경사항

- 투표의 모든 상태를 초기화하는 기능이 추가됩니다. 이때 후보 등록 정보는 초기화 되지 않으나, 회원들의 투표 정보와 회원이 투표 결과보기 상태 여부를 초기화합니다.
- VoteResultMemberRepository 에서 delete 관련 메서드 추가
- reset API 추가

### Issue Link - #152

Nine-JH committed:
[Feat] 투표 초기화 기능 추가

motivation:
- 투표의 모든 상태를 초기화하는 기능이 추가됩니다. 이때 후보 등록 정보는 초기화 되지 않으나, 회원들의 투표 정보와 회원이 투표 결과보기 상태 여부를 초기화합니다.

modification:
- VoteResultMemberRepository 에서 delete 관련 메서드 추가
- reset API 추가

Nine-JH committed:
Merge branch 'develop' into feature/#152-reset-vote-status

Nine-JH committed:
[Fix] 컴파일 에러 수정

motivation:
- Return 값이 없던 Controller을 수정했습니다.

modification:
- `VoteControllerImpl.resetVote()` 컴파일 에러 수정

laigasus commented on app/src/main/java/fc/be/app/domain/vote/controller/dto/request/CandidateDeleteApiRequest.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.List;
+
+public record CandidateDeleteApiRequest(
+        @NotNull List<Long> candidateIds
+) {
```
이거 후보지 id 최소 하나라도 있어야 되는 것이라면 
`@Size(min = 1)` 를 붙일 것 같아요!

laigasus commented on app/src/main/java/fc/be/app/domain/vote/service/service/VoteManageService.java:
```diff
@@ -163,8 +173,49 @@ public void deleteVote(Long voteId, Long memberId) {
     }
+
+    public void deleteCandidates(CandidateDeleteRequest request) {
+        final Member requestMember = memberRepository.findById(request.memberId())
+                .orElseThrow(() -> new MemberException(MEMBER_NOT_FOUND));
+
+        Vote vote = voteRepository.findById(request.voteId())
+                .orElseThrow(() -> new VoteException(VoteErrorCode.VOTE_NOT_FOUND));
```
개인적으로 Vote 쿼리 한번 수행해서 Member를 가져오는 것은 어떠신지요?

Controller에서 `@AuthenticationPrincipal` 을 통해서 인증된 회원 정보를 가져오니까요

검증된 회원 정보에 한해서 voteRepository에 자원 접근쿼리 한번 이면 충분하다고 생각합니다.

```java
Vote vote = voteRepository.findByIdWithMemberAndCandidates(request.voteId(), request.memberId())
            .orElseThrow(() -> new VoteException(VoteErrorCode.VOTE_NOT_FOUND));
```

이후 로직에서 회원정보 필요하면 vote 내 회원 꺼내 쓸 수 있기도 하구요

Nine-JH replied:
사실 이게 뒤죽박죽이 되어버린 케이스입니다. 원래 vote에 member이 연관관계가 아닌 member_id와 같은 형식으로 되어있어서 해당 방식을 취했었는데 지금 상황에서는 굳이 해당 방식을 고수할 필요가 없겠네요 😄 조언 감사합니다

laigasus commented on app/src/main/java/fc/be/app/domain/vote/service/service/VoteManageService.java:
```diff
@@ -163,8 +173,49 @@ public void deleteVote(Long voteId, Long memberId) {
+    }
+
+    public void resetVote(Long voteId, Long memberId) {
+        final Member requestMember = memberRepository.findById(memberId)
+                .orElseThrow(() -> new MemberException(MEMBER_NOT_FOUND));
+
+        Vote vote = voteRepository.findById(voteId)
+                .orElseThrow(() -> new VoteException(VoteErrorCode.VOTE_NOT_FOUND));
```
> 개인적으로 Vote 쿼리 한번 수행해서 Member를 가져오는 것은 어떠신지요?
> 
> Controller에서 `@AuthenticationPrincipal` 을 통해서 인증된 회원 정보를 가져오니까요
> 
> 검증된 회원 정보에 한해서 voteRepository에 자원 접근쿼리 한번 이면 충분하다고 생각합니다.
> 
> ```java
> Vote vote = voteRepository.findByIdWithMemberAndCandidates(request.voteId(), request.memberId())
>             .orElseThrow(() -> new VoteException(VoteErrorCode.VOTE_NOT_FOUND));
> ```
> 
> 이후 로직에서 회원정보 필요하면 vote 내 회원 꺼내 쓸 수 있기도 하구요

2

laigasus commented on app/src/main/java/fc/be/app/domain/vote/service/service/VoteManageService.java:
```diff
@@ -163,8 +173,49 @@ public void deleteVote(Long voteId, Long memberId) {
+        this.resetResultMode(voteId, memberId);
+    }
+
+    public void changeToResultMode(Long spaceId, Long voteId, Long memberId) {
+        if (voteResultMemberRepository.existsByMemberIdAndVoteId(memberId, voteId)) {
+            return;
+        }
+        voteResultMemberRepository
+                .save(VoteResultMember.of(memberId, voteId, spaceId));
+    }
```
이거는 `@Modifying` 을 이용하시면 한번의 쿼리로 수행하실 수 있을 것 같아요.
AI 발 코드 함 올려봅니다.

```java
// VoteResultMemberRepository.java
@Modifying
@Query("INSERT INTO VoteResultMember (memberId, voteId, spaceId) SELECT :memberId, :voteId, :spaceId WHERE NOT EXISTS (SELECT 1 FROM VoteResultMember WHERE memberId = :memberId AND voteId = :voteId)")
void saveIfNotExists(@Param("memberId") Long memberId, @Param("voteId") Long voteId, @Param("spaceId") Long spaceId);
```

Nine-JH replied:
<img width="359" alt="image" src="https://github.com/Strong-Potato/TripVote-BE/assets/139187207/19d6af9d-d4e8-4090-99e2-2de026035165">

이거 서브쿼리를 썼을때 빨간줄이 생기길래 컴파일 에러인가 싶었는데 되는 구문이었군요 😠 
다시 작성해보겠습니다 😄

laigasus commented on app/src/main/java/fc/be/app/domain/vote/service/service/VoteManageService.java:
```diff
@@ -41,16 +46,26 @@ public class VoteManageService {
+        this.votedMemberRepository = votedMemberRepository;
+        this.voteResultMemberRepository = voteResultMemberRepository;
     }
 
     @Transactional
```
클래스 위에 `@Transactional` 있던데, 따로 넣어야 하는 옵션이 있을까요? 🤔

Nine-JH replied:
먼저 해당 service는 CUD만을 다루는 서비스이기 때문에 readOnly는 필요없습니다.
추가적으로 vote의 경우 다른 entity로 전이되는 케이스가 없어서 isolation과 propagation이 없어도 무방하다고 판단을 했습니다.
candidate의 경우에는 같은 도메인 관계기 때문에 더욱 필요없다고 생각했고요

laigasus replied:
@Nine-JH 답변 감사합니다

`@Transactional` 이렇게 두번 사용하셨길래용. 일케 해야 올바르게 동작하는 기능일까요?
https://github.com/Strong-Potato/TripVote-BE/blob/148db98cfd7fb4e7cb440a24aee64de5f042cb5e/app/src/main/java/fc/be/app/domain/vote/service/service/VoteManageService.java#L39-L73

Nine-JH replied:
아 저건 메서드에 작성이 된거군요 ㅋㅋㅋㅋㅋㅋㅋ
죄송합니다 착각했어요

laigasus commented on app/src/main/java/fc/be/app/domain/vote/service/service/VoteInfoQueryService.java:
```diff
@@ -40,16 +39,19 @@ public class VoteInfoQueryService {
+    private final VoteManageService voteManageService;
 
     public VoteInfoQueryService(VoteRepository voteRepository,
                                 SpaceRepository spaceRepository,
                                 MemberRepository memberRepository,
-                                VoteResultMemberRepository voteResultMemberRepository
+                                VoteResultMemberRepository voteResultMemberRepository,
+                                VoteManageService voteManageService
     ) {
         this.voteRepository = voteRepository;
         this.spaceRepository = spaceRepository;
         this.memberRepository = memberRepository;
         this.voteResultMemberRepository = voteResultMemberRepository;
+        this.voteManageService = voteManageService;
     }
```
lombok을 경계하셔서 이렇게 구현하신 것 같은데 ㅋㅋㅋ

리팩터링때 '통일' 해보는 것은 어떠신지요 🤗

Nine-JH replied:
그것도 좋죠 😄

laigasus reviewed (approved):
수고하셨습니다 🙇
남은 기간동안 ㅎㅇㅌ 이에요 💪

소소한 코멘트 남겼습니다!

service단에서 간간히 db 조회 횟수가 많은 것들이 보이네요.
리팩터링때 줄여나가 보아요 😊

Nine-JH committed:
[Feat] PR 코드 리팩토링

motivation:
- PR(#153)에서 받은 comment 를 통해 몇몇 코드를 변경하였습니다.

modification:
- `CandidateAddApiRequest.java` `@Size` 추가
- `VoteManageService.java`의 메서드 레벨에 존재하는 `@Transactional` 제거

Dr-KoKo commented on app/src/main/java/fc/be/app/domain/vote/entity/VotedMember.java:
```diff
@@ -18,7 +18,7 @@ public class VotedMember {
+    @ManyToOne(cascade = CascadeType.REMOVE)
     @JoinColumn(name = "candidate_id")
     @Comment("투표한 장소")
     private Candidate candidate;
```
지금대로면 VotedMember가 삭제될 때 `Candidate`랑 `Vote`가 삭제되는 것으로 알고 있는데 cascade 설정이 Candidate쪽에 되어 있어야 하는 거 아닌가요? (제가 잘못 알고 있는 것일 수도?)

Nine-JH replied:
아 이거 테스트 목적으로 작성해봤는데 그대로 딸려들어갔네요 😓  제거하도록 하겠습니다

Dr-KoKo commented on app/src/main/java/fc/be/app/domain/vote/entity/VotedMember.java:
```diff
@@ -28,7 +28,7 @@ public class VotedMember {
     private Member member;
 
-    @ManyToOne
+    @ManyToOne(cascade = CascadeType.REMOVE)
     @JoinColumn(name = "vote_id")
     @Comment("투표 id")
     private Vote vote;
```
>지금대로면 VotedMember가 삭제될 때 `Candidate`랑 `Vote`가 삭제되는 것으로 알고 있는데 cascade 설정이 Candidate쪽에 되어 있어야 하는 거 아닌가요? (제가 잘못 알고 있는 것일 수도?)

Dr-KoKo reviewed (approved):
수고하셨습니다. 문서화까지 👍

Nine-JH committed:
[Feat] PR 코드 리팩토링

motivation:
- PR(#153)에서 받은 comment 를 통해 몇몇 코드를 변경하였습니다.

modification:
- `cascade` option 제거
